components/layout/query.rs

```diff
@@ -527,21 +530,58 @@ impl FragmentBorderBoxIterator for UnioningFragmentScrollAreaIterator {
 // https://drafts.csswg.org/cssom-view/#extensions-to-the-htmlelement-interface
 impl FragmentBorderBoxIterator for ParentOffsetBorderBoxIterator {
     fn process(&mut self, fragment: &Fragment, level: i32, border_box: &Rect<Au>) {
+        match self.node_offset_box {
+            Some(_) => {  }, // We've already found the node, so we already have its parent.
+            None => {
+                // Remove all nodes at this level or higher, as they can't be parents of this node.
+                self.parent_nodes.truncate(level as usize);
+                assert!(self.parent_nodes.len() == level as usize);
+            },
+        }
+
         if fragment.node == self.node_address {
             // Found the fragment in the flow tree that matches the
             // DOM node being looked for.
-            self.has_found_node = true;
-            self.node_border_box = *border_box;
+            self.has_processed_node = true;
+            self.node_offset_box = Some(NodeOffsetBoxInfo {
+                offset: border_box.origin,
+                rectangle: *border_box,
+            });
 
             // offsetParent returns null if the node is fixed.
             if fragment.style.get_box().position == computed_values::position::T::fixed {
                 self.parent_nodes.clear();
             }
-        } else if level > self.last_level {
+        } else if fragment.contains_node(self.node_address) {
+            // Found a fragment in the flow tree that matches a child
+            // of the DOM node being looked for.
+            // Since we haven't found an exact match, the node must be
+            // inline. Hopefully.
+            match self.node_offset_box {
+                Some(NodeOffsetBoxInfo { ref mut rectangle, .. }) => {
+                    *rectangle = rectangle.union(border_box);
```

Comment:
Text that starts in the middle of a line and has a line break will have boxes further to the left than the first one, so they'll change the `origin`. I imagine right-to-left text will do so as well.

The offset properties are very confusing. The position `(offsetLeft, offsetTop)` doesn't actually correspond to the upper-left corner of a rectangle with dimensions `offsetWidth x offsetHeight`.

Come to think of it, right-to-left text might mess up the result of `offsetLeft` from this code. Then again, off the top of my head it seems like the "messed-up" results I'm imagining, unintuitive though they are, may actually be the correct ones according the spec. I'll have to look into that.